javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
   }
 
   /* `KeySetView` is a public class in the JDK API. The result of
-   * `ConcurrentHashMap.keySet()` must be statically typed as a `KeySetView`,
-   * hence the existence of this class, although it forwards all its operations
-   * to the inner key set.
+   * `ConcurrentHashMap.keySet()` must be statically typed as a `KeySetView`.
    */
-  class KeySetView[K, V] private[ConcurrentHashMap] (inner: Set[K])
+  class KeySetView[K, V] private[ConcurrentHashMap] (innerMap: InnerHashMap[K, V], defaultValue: V)
       extends Set[K] with Serializable {
 
-    def contains(o: Any): Boolean = inner.contains(o)
+    def contains(o: Any): Boolean = innerMap.containsKey(o)
 
-    def remove(o: Any): Boolean = inner.remove(o)
+    def remove(o: Any): Boolean = innerMap.remove(o) != null
 
-    def iterator(): Iterator[K] = inner.iterator()
+    def iterator(): Iterator[K] = innerMap.keySet().iterator()
 
-    def size(): Int = inner.size()
+    def size(): Int = innerMap.size()
 
-    def isEmpty(): Boolean = inner.isEmpty()
+    def isEmpty(): Boolean = innerMap.isEmpty()
 
-    def toArray(): Array[AnyRef] = inner.toArray()
+    def toArray(): Array[AnyRef] = innerMap.keySet().toArray()
 
-    def toArray[T <: AnyRef](a: Array[T]): Array[T] = inner.toArray[T](a)
+    def toArray[T <: AnyRef](a: Array[T]): Array[T] = innerMap.keySet().toArray(a)
 
-    def add(e: K): Boolean = inner.add(e)
+    def add(e: K): Boolean = {
+      if (e == null || defaultValue == null) {
+        throw new UnsupportedOperationException()
+      }
+      innerMap.putIfAbsent(e, defaultValue) == null
+    }
 
-    def containsAll(c: Collection[_]): Boolean = inner.containsAll(c)
+    def containsAll(c: Collection[_]): Boolean = innerMap.keySet().containsAll(c)
 
-    def addAll(c: Collection[_ <: K]): Boolean = inner.addAll(c)
+    def addAll(c: Collection[_ <: K]): Boolean = {
+      if (c == null || defaultValue == null) {
+        throw new UnsupportedOperationException()
+      }
+      val iter = c.iterator()
+      var addedAny: Boolean = false
+      while (iter.hasNext()) {
+        val nextKey = iter.next()
+        if (!innerMap.containsKey(nextKey)) {
+          innerMap.put(nextKey, defaultValue)
+          addedAny = true
+        }
```

Comment:
`addedAny ||= innerMap.putIfAbsent(iter.next(), defaultValue)`?

Reply:
Hmm, I tried the below 

```scala
      while (iter.hasNext()) {
        val nextKey = iter.next()
        addedAny ||= innerMap.putIfAbsent(nextKey, defaultValue) == null
      }
```
but this made some tests failing.

Reply:
Potentially the same issue than for `add`.

Reply:
The below tests are failing, even after fixed `NullRejectingHashMap.putIfAbsent`.

```
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.simpleSum failed: java.lang.AssertionError: expected:<268> but was:<42>, took 0.001999872 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.toStringCollectionCustomClass failed: java.lang.AssertionError: result '[Custom(A,1)]' not in expected set 'HashSet([Custom(b,2), Custom(A,1), Custom(C,3)], [Custom(b,2), Custom(C,3), Custom(A,1)], [Custom(C,3), Custom(b,2), Custom(A,1)], [Custom(C,3), Custom(A,1), Custom(b,2)], [Custom(A,1), Custom(b,2), Custom(C,3)], [Custom(A,1), Custom(C,3), Custom(b,2)])', took 0.004 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.toStringCollectionDoubleHasCommaSpace failed: java.lang.AssertionError: result '[88.42]' not in expected set 'HashSet([60.173, 88.42, -23.36], [88.42, -23.36, 60.173], [-23.36, 60.173, 88.42], [60.173, -23.36, 88.42], [88.42, 60.173, -23.36], [-23.36, 88.42, 60.173])', took 0.001000192 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.removeIf failed: java.lang.AssertionError: expected:<8> but was:<1>, took 0 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.iterator failed: java.lang.AssertionError: expected:<3> but was:<1>
```

Reply:
Why are they failing? Also, I just realize that using `add` might even be a better option. Like here: https://github.com/scala-js/scala-js/blob/5cc821bd51ba49d7ee33e5a2dffa44636864da41/javalib/src/main/scala/java/util/HashSet.scala#L57-L63

Reply:
Ah, that one using `add` worked well.

Reply:
Ah, that's my bad :) The problem was short-circuiting behavior of `||`.